core/src/main/scala/cats/data/ContT.scala

```diff
@@ -0,0 +1,109 @@
+package cats
+package data
+
+/**
+ * This is a continuation transformer based on the ContT in
+ * the Haskell package Control.Monad.Cont
+ *
+ * This is reasonably straight-forward except that to make
+ * a tailRecM implementation we leverage the Defer type class to
+ * obtain stack-safety.
+ */
+sealed abstract class ContT[M[_], A, B] extends Serializable { self =>
+  final def run: (B => M[A]) => M[A] = runAndThen
+  protected def runAndThen: AndThen[B => M[A], M[A]]
+
+  def map[C](fn: B => C)(implicit M: Defer[M]): ContT[M, A, C] = {
+    // allocate/pattern match once
+    val fnAndThen = AndThen(fn)
+    ContT { fn2 =>
+      val cb = fnAndThen.andThen(fn2)
+      M.defer(run(cb))
+    }
+  }
+
+  /**
+   * c.mapCont(f).run(g) == f(c.run(g))
+   */
+  def mapCont(fn: M[A] => M[A]): ContT[M, A, B] =
+    // Use later here to avoid forcing run
+    ContT.later(runAndThen.andThen(fn))
+
+  /**
+   * cont.withCont(f).run(cb) == cont.run(f(cb))
+   */
+  def withCont[C](fn: (C => M[A]) => B => M[A]): ContT[M, A, C] =
+    // lazy to avoid forcing run
+    ContT.later(AndThen(fn).andThen(runAndThen))
+
+  def flatMap[C](fn: B => ContT[M, A, C])(implicit M: Defer[M]): ContT[M, A, C] = {
+    // allocate/pattern match once
+    val fnAndThen = AndThen(fn)
+    ContT[M, A, C] { fn2 =>
+      val contRun: ContT[M, A, C] => M[A] = (_.run(fn2))
+      val fn3: B => M[A] = fnAndThen.andThen(contRun)
+      M.defer(run(fn3))
+    }
+  }
+}
+
+object ContT {
+
+  // Note, we only have two instances of ContT in order to be gentle on the JVM JIT
+  // which treats classes with more than two subclasses differently
+
+  private case class FromFn[M[_], A, B](runAndThen: AndThen[B => M[A], M[A]]) extends ContT[M, A, B]
+
+  private case class DeferCont[M[_], A, B](next: () => ContT[M, A, B]) extends ContT[M, A, B] {
+    @annotation.tailrec
+    private def loop(n: () => ContT[M, A, B]): ContT[M, A, B] =
+      n() match {
+        case DeferCont(n) => loop(n)
+        case notDefer => notDefer
+      }
+
+    lazy val runAndThen: AndThen[B => M[A], M[A]] = loop(next).runAndThen
+  }
+
+  def pure[M[_], A, B](b: B): ContT[M, A, B] =
+    apply { cb => cb(b) }
+
+  def apply[M[_], A, B](fn: (B => M[A]) => M[A]): ContT[M, A, B] =
+    FromFn(AndThen(fn))
```

Comment:
I'd also like to see an `applyPure` or something like that defined as:

```scala
def apply[M[_]: Applicative, A, B](fn: (B => A) => A): ContT[M, A, B] =
  apply[M, A, B](_(fn.andThen(_.pure)).pure)
```

Or thereabouts…

Reply:
actually, I don't think we can do that I think we need Applicative and Comonad (we need to go `B => M[A]` to `B => A` to call `fn`.

Reply:
Oh I see what you mean. That's definitely unfortunate. I'm not sure the function would be useful at all requiring `Comonad`